### DISCUSSION TITLE
Связь материалов внутри раздела

### DISCUSSION THREAD

**Discussion by pepelsbey** [doka-guide/content#753] [Разное]
У нас была задача «[Продумать систему заголовков и вывод оглавления](https://github.com/Y-Doka/content/issues/348)», по следам которой было обсуждение и решение в виде задачи «[Описать структуру материалов в разделах](https://github.com/Y-Doka/content/issues/738)».

Мы решили, что хотим связать материалы и выводить их в индексе полезнее и удобнее. Есть парочка черновых структур:

- [Алёна про вёрстку](https://disk.yandex.ru/i/IsdexK6jR5VwOQ)
- [Коля про JS и тулинг](https://www.notion.so/praktikum/d82a666a662844eb8e09022dfe02f83f)

Эта дискуссия посвящена тому, как это реализовать технически в контенте.

Чего хотелось бы добиться:

- Обозначить связь родитель-ребёнок (например, в контентной мете)
- Хранить порядок вывода статей в группе (не в алфавитном порядке)
- Желательно, не создавать дополнительных файлов и конфигов

Давайте придумаем, как. Некоторые мысли уже [прозвучали раньше](https://github.com/Y-Doka/content/issues/348#issuecomment-843621071), давайте продолжим.

**Comment by pepelsbey** [comment]
# Задача

Предположим, у нас есть четыре статьи, одна из которых корневая, а три вложенных:

```
html/a/index.md
html/dl/index.md
html/tags/index.md ← корневая
html/video/index.md
```

Их связи никак не обозначены на файловой структуре. Здесь я уже предусмотрел вариант, в котором мы откажемся от категорий article/doka, [см. ишью об этом](https://github.com/Y-Doka/content/issues/737).

Мы хотим вывести эти статьи в индексе в таком порядке:

- Теги
  - video
  - dl
  - a

Не спрашивайте, почему именно в таком. Хотим, надо :)

# Решение

Во вложенных файлах обозначить родителя:

```diff
  ---
  title: video
  authors:
    - realetive
+ parent: tags
  ---
```
```diff
  ---
  title: dl
  authors:
    - ezhkov
+ parent: tags
  ---
```
```diff
  ---
  title: a
  authors:
    - solarrust
+ parent: tags
  ---
```

Дальше, в родительском (корневом) файле нужно обозначить порядок:

```diff
  ---
  title: Теги
  authors:
    - solarrust
+ children:
+   - video
+   - dl
+   - a
  ---
```

В таком случае, при выводе индекса мы сможем вывести детей под родителем в нужном порядке.

Стоит также предусмотреть защиту от случайностей: если в корневом материале не перечислен какой-то из вложенных, он просто попадает в конец списка по алфавиту.

**Comment by pepelsbey** [comment]
@nlopin что скажешь? :)

**Comment by solarrust** [comment]
Само предложение со связкой материалов в мета-шапках — норм. Но немного смущает двойная связь в том смысле, что придётся каждый раз при написании статьи идти и искать файл-родитель и думать над тем, в каком порядке выводить материалы. Хотя я понимаю обоснованность этого решения: возможность сортировки материалов внутри раздела. 
Если можно будет остаться в рамках меты, но уйти от двойных связей — будет круто.

Ещё можно попросить показать вариант когда материал относится к двум-трём разделам? Кажется, что на этот случай стоит сделать `parent` списком

**Reply by pepelsbey** [reply]
@solarrust да, можно сделать и к нескольким родителям, правда я не уверен, что это не запутает: ты всё же ожидаешь, что в индексе выводятся уникальные статьи, без повторов.

```diff
  ---
  title: video
  authors:
    - realetive
+ parent:
    - tags
    - media
  ---
```

**Reply by solarrust** [reply]
Но такие ситуации у нас точно будут. Смотри, статья про `<form>` она условно будет относится и к разделу «Теги» и к разделу «Формы».

**Reply by pepelsbey** [reply]
Хотя да, с формами это выглядит не так страшно. Но стоит не злоупотреблять этой возможностью

**Comment by nlopin** [comment]
Я думал, как бы сделать так, чтобы не нужно было бы указывать значения и там, и там. Раз для нас важен порядок, то указывать связь `children → parent` в дочернем элементе не имеет смысла — дочерний элемент не владеет всей информацией, чтобы выбрать позицию.

Полной информацией о своих дочерних узлах обладает только родительский элемент. Предлагаю указывать список дочерних узлов только в нём, а дочерние оставить в покое:

```diff
  ---
  title: Теги
  authors:
    - solarrust
+ children:
+   - video
+   - dl
+   - a
  ---
```

Мы можем написать скриптик, который выводит все статьи без родителей. Такой подход нам может пригодиться, если в какой-то момент мы захотим публиковать не одиночные статьи, а разделы.

**Reply by pepelsbey** [reply]
Действительно, кажется так будет проще :)

**Comment by solarrust** [comment]
Поинт: самому разделу тоже нужен индекс, чтобы в нём указывать порядок вывода разделов. Чтобы не в алфавитном, а в логическом порядке расположить их на индексной странице.

**Comment by pepelsbey** [comment]
Продолжаем думать про решение!

Мы отказались от корневых статей и будем делать вводные статьи вложенными в категории, они будут идти первыми.

Поэтому вот идея №2: положить в папку раздела индексный файл, который описывает его структуру.

Например, представим файл html/index.md

```markdown
---
- name: 'Теги'
  items:
    - video
    - dl
    - a
- name: 'Формы'
  items:
  - input
  - textarea
---
```

В этом случае у нас на странице выведется:

# HTML
## Теги
- `<video>`
- `<dl>`
- `<a>`
## Формы
- `<input>`
- `<textarea>`

**Reply by solarrust** [reply]
Мне норм такой формат. Но хочу напомнить про вывод неотсортированный статей и ворнинг на то, что статья не упомянута в индексе. Можно при ПРе

**Comment by pepelsbey** [comment]
FYI, сделал черновики ПР с индексами. Переходим к внедрению :)

- [CSS](https://github.com/doka-guide/content/pull/1115)
- [HTML](https://github.com/doka-guide/content/pull/1116)
- [JavaScript](https://github.com/doka-guide/content/pull/1117)
- [Инструменты ](https://github.com/doka-guide/content/pull/1118)

**Comment by nlopin** [comment]
Итоговый формат чуть изменился, добавилось описание раздела. [Пример структуры раздела HTML](https://github.com/doka-guide/content/blob/main/html/index.md)
